client/src/components/TimeoutDialog.vue

```diff
@@ -1,104 +1,105 @@
+    });
 
-      // reset no-action timer
-      this.setActionTimeout(false);
-      this.resetActionTimer();
-    },
-    logout() {
-      this.minutes = 0;
-      this.seconds = 0;
-    },
-    reload() {
-      this.$router.go();
-    },
-    decrement() {
-      if (this.show) {
-        setTimeout(() => {
-          this.seconds -= 1;
+    const lastApiRequestTime = computed(() => store.state.lastApiRequestTime);
 
-          if (this.minutes <= 0 && this.seconds <= 0) {
-            djangoRest.logout();
-            return;
-          }
+    const reset = () => {
+      // Send a request to refresh the server token
+      // Not awaited since we don't actually care about the result
+      djangoRest.projects();
 
-          if (this.seconds === 0) {
-            this.minutes -= 1;
-            this.seconds = initSeconds;
-          }
+      // reset dialog
+      show.value = false;
+      idleWarningTriggered.value = false;
+    };
+    const logout = async () => {
+      await djangoRest.logout();
+      // This will redirect to the login page
+      await djangoRest.login();
+    };
 
-          this.decrement();
-        }, 1000);
+    // Watch for an absence of user interaction
+    const idle = new IdleJS({
+      idle: idleTimeout - warningDuration,
+      onIdle() {
+        if (!show.value) {
+          idleWarningTriggered.value = true;
+          idleStartTime.value = Date.now();
+        }
+      },
+    });
+    idle.start();
+
+    // This function calls itself after 1 second to check if the session has expired and to keep
+    // the countdown timer up to date.
+    const updateCountdown = () => {
+      const now = Date.now();
+      console.log('updating countdown', now);
+      const sessionTimeRemaining = lastApiRequestTime.value + sessionTimeout - now;
+      if (idleWarningTriggered.value) {
+        // If the user is idle, we also need to consider the idle warning
+        const idleTimeRemaining = idleStartTime.value + warningDuration - now;
+        timeRemaining.value = Math.min(sessionTimeRemaining, idleTimeRemaining);
+      } else {
+        timeRemaining.value = sessionTimeRemaining;
+      }
+      // The timer has expired, log out
+      if (timeRemaining.value <= 0) {
+        logout();
       }
-    },
+      // Show the warning if the time remaining is getting close to 0
+      show.value = timeRemaining.value < warningDuration;
+      setTimeout(updateCountdown, 1000);
+    };
+    updateCountdown();
+    // TODO when the webpack dev server reloads, it doesn't stop this setTimeout loop.
+    // The component that the old loop was servicing no longer exists so it doesn't actually matter,
+    // but it would be nice to garbage collect it.
+
+    return {
+      show,
+      idleWarningTriggered,
+      timeRemaining,
+      timeRemainingStr,
+      reset,
+      logout,
+      sessionTimeout,
+      idleTimeout,
+    };
   },
-};
+  onIdle() {
+    // TODO why this does not trigger
```

Comment:
This seems like a blocker; what is this function meant to do and is it still not happening?

Reply:
This was the idle-vue hook that I could not get to trigger. I opted to replace idle-vue with idle-js, which was more straightforward to set up and didn't involve any Vue magic.